.github/workflows/build-frontend.yml

```diff
@@ -13,29 +13,58 @@ on:
       - "!.github/workflows/stellar-dominion-service*.yml"
 
 jobs:
+  # https://github.com/dorny/paths-filter
+  analyze-changes:
+    runs-on: ubuntu-latest
+    outputs:
+      user-dashboard: ${{ steps.filter.outputs.user-dashboard }}
+      stellar-dominion: ${{ steps.filter.outputs.stellar-dominion }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dorny/paths-filter@v3
+        id: filter
+        with:
+          base: ${{ github.ref_name }}
+          # https://github.com/dorny/paths-filter?tab=readme-ov-file#conditional-execution
+          filters: |
+            user-dashboard:
+              - 'website/user-dashboard/**'
+            stellar-dominion:
+              - 'website/stellar-dominion/**'
   build-user-dashboard:
+    needs: [analyze-changes]
+    if: ${{ needs.analyze-changes.outputs.user-dashboard == 'true' }}
     uses: ./.github/workflows/user-dashboard-build-and-push.yml
     secrets:
       dockerhub-username: ${{ secrets.DOCKERHUB_USERNAME }}
       dockerhub-token: ${{ secrets.DOCKERHUB_TOKEN }}
   build-stellar-dominion:
+    needs: [analyze-changes]
+    if: ${{ needs.analyze-changes.outputs.stellar-dominion == 'true' }}
     uses: ./.github/workflows/stellar-dominion-build-and-push.yml
     secrets:
       dockerhub-username: ${{ secrets.DOCKERHUB_USERNAME }}
       dockerhub-token: ${{ secrets.DOCKERHUB_TOKEN }}
   persist-service-tags:
     runs-on: ubuntu-latest
-    if: ${{ github.ref == 'refs/heads/master' }}
-    needs: [build-user-dashboard, build-stellar-dominion]
+    needs: [analyze-changes, build-user-dashboard, build-stellar-dominion]
+    # https://github.com/actions/runner/issues/491
```

Comment:
This was an important issue:
* `needs` is supposed to indicate that the job requires some output from another job: in this case if the job is skipped then it makes sense to automatically skip the jobs that require it.
* on the other hand `needs` is the only way to ensure sequential execution of different steps: in this context it might not make sense to automatically skip it.

This [comment](https://github.com/actions/runner/issues/491#issuecomment-846374956) explains it nicely.